Sources/Tracking/Background Queue/Queue.swift

```diff
@@ -46,55 +46,87 @@ public class CioQueue: Queue {
+        sdkConfigStore.config.backgroundQueueSecondsDelay
+    }
 
     init(
         siteId: SiteId,
         storage: QueueStorage,
         runRequest: QueueRunRequest,
         jsonAdapter: JsonAdapter,
         logger: Logger,
-        sdkConfigStore: SdkConfigStore
+        sdkConfigStore: SdkConfigStore,
+        queueTimer: SingleScheduleTimer
     ) {
         self.storage = storage
         self.siteId = siteId
         self.runRequest = runRequest
         self.jsonAdapter = jsonAdapter
         self.logger = logger
         self.sdkConfigStore = sdkConfigStore
+        self.queueTimer = queueTimer
     }
 
     public func addTask<T: Codable>(type: String, data: T) -> (success: Bool, queueStatus: QueueStatus) {
+        logger.info("adding queue task \(type)")
+
         guard let data = jsonAdapter.toJson(data, encoder: nil) else {
+            logger.error("fail adding queue task, json encoding fail.")
+
             return (success: false,
                     queueStatus: QueueStatus(queueId: siteId, numTasksInQueue: storage.getInventory().count))
         }
 
+        logger.debug("added queue task data \(data.string ?? "")")
+
         let addTaskResult = storage.create(type: type, data: data)
         processQueueStatus(addTaskResult.queueStatus)
 
         return addTaskResult
     }
 
     public func run(onComplete: @escaping () -> Void) {
-        logger.verbose("Manually running background queue")
+        logger.info("queue run request sent")
 
         runRequest.start(onComplete: onComplete)
     }
 
+    /// We determine the queue needs to run by (1) if there are many tasks in the queue
+    /// (2) we schedule tasks to run sometime in the future.
+    /// It is by grouping more then 1 task to run at a time in the queue that will save
+    /// the device battery life so we try to do that when we can.
     private func processQueueStatus(_ status: QueueStatus) {
-        logger.verbose("Processing queue status \(status).")
+        logger.debug("processing queue status \(status).")
         let isManyTasksInQueue = status.numTasksInQueue >= sdkConfigStore.config.backgroundQueueMinNumberOfTasks
 
-        let runQueue = isManyTasksInQueue
+        if isManyTasksInQueue {
+            logger.info("queue met criteria to run automatically")
 
-        if runQueue {
-            logger.verbose("Automatically running background queue")
+            // cancel timer if one running since we will run the queue now
+            queueTimer.cancel()
 
             // not using [weak self] to assert that the queue will complete and callback once started.
             // this might keep this class in memory and not get garbage collected once customer is done using it
             // but it will get released once the queue is done running.
             runRequest.start {
-                self.logger.verbose("Automatic running background queue completed")
+                self.logger.info("queue completed all tasks")
```

Comment:
thinking about it some more, what makes sense to me is:
- try to send all messages now
- if any fail with a 40x, ignore them
- if we hit a 50x, halt and schedule another run